pkg/container-runtime/configurator.go

```diff
@@ -0,0 +1,305 @@
+		return err
+	}
+
+	return nil
+}
+
+func (c ContainerdRuntimeConfigurator) configHostsFile(registryIP, domain string, ips []net.IP) error {
+	// add registry ip to "/etc/hosts"
+	hostsLine := registryIP + " " + domain
+	writeToHostsCmd := fmt.Sprintf(RemoteAddEtcHosts, hostsLine, hostsLine)
+
+	f := func(host net.IP) error {
+		err := c.infraDriver.CmdAsync(host, writeToHostsCmd)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return concurrencyExecute(f, ips)
+}
+
+func (c ContainerdRuntimeConfigurator) configRegistryCert(domain, port string, caFile string, ips []net.IP) error {
+	// copy ca cert to "/etc/containerd/certs.d/${domain}:${port}/ca.crt
+	src := filepath.Join(c.rootfs, "certs", caFile)
+	dest := filepath.Join(ContainerdCertDir, domain+":"+port, caFile)
+
+	if !os.IsFileExist(src) {
+		return nil
+	}
+
+	f := func(host net.IP) error {
+		err := c.infraDriver.Copy(host, src, dest)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return concurrencyExecute(f, ips)
+}
+
+func (c ContainerdRuntimeConfigurator) configKubeletAuthInfo(username, password string, url string, ips []net.IP) error {
+	if username == "" || password == "" {
+		return nil
+	}
+
+	configAuthCmd := fmt.Sprintf(NerdctlLoginCommand, username, password, url)
+
+	f := func(host net.IP) error {
+		err := c.infraDriver.CmdAsync(host, configAuthCmd)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return concurrencyExecute(f, ips)
+}
+
+func (c ContainerdRuntimeConfigurator) configContainerd(endpoint string, ips []net.IP) error {
+	configContainerdCmd := []string{}
+	configFile := filepath.Join(c.rootfs, "etc/dump-config.toml")
+
+	if endpoint != DefaultEndpoint {
+		configContainerdCmd = append(configContainerdCmd,
+			fmt.Sprintf(ModifyFileCommand, DefaultEndpoint, endpoint, configFile))
+	}
+
+	//TODO config registry basic auth to dump-config.toml
+	//if username != "" && password != "" {
+	//
+	//}
+
+	// containerd --config ${dump-config.toml} config dump > /etc/containerd/config.toml
+	// systemctl restart containerd
+	modifyContainerdDaemonFileCmd := fmt.Sprintf(DumpConfigContainerd, configFile)
+	restartDockerCommand := fmt.Sprintf(RestartServiceCommand, "containerd")
+	configContainerdCmd = append(configContainerdCmd, modifyContainerdDaemonFileCmd, restartDockerCommand)
+
+	f := func(host net.IP) error {
+		err := c.infraDriver.CmdAsync(host, strings.Join(configContainerdCmd, " & "))
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return concurrencyExecute(f, ips)
+}
+
+func NewContainerdRuntimeDriver(rootfs string, infraDriver infradriver.InfraDriver) Configurator {
+	return ContainerdRuntimeConfigurator{
+		rootfs:      rootfs,
+		infraDriver: infraDriver,
+	}
+}
+
+func concurrencyExecute(f func(host net.IP) error, ips []net.IP) error {
+	eg, _ := errgroup.WithContext(context.Background())
+	for _, ip := range ips {
+		host := ip
+		eg.Go(func() error {
+			err := f(host)
+			if err != nil {
+				return err
```

Comment:
Please wrap this error. I think the error message should carry the following informtion:
* host 
* command
* error information